SoapHandlers with EJB + Servlet based Webservices are flaky

# Description #
----------

Summary: When EJBs and Servlet based Webservice Endpoints are used, SoapHandlers are configured and EJBs are injected into the Servlet based endpoint, the SoapHandlers are not always configured/instantiated correctly. This results in EJB based Webservices without SoapHandlers.

## Steps to reproduce

1. Startpoint: Get Payara 4.1.1.164 and expand it
2. Start payaradomain: `bin/asadmin start-domain payaradomain`
3. Clone https://github.com/matthiasblaesing/SoapHandlerEJBAndServlet.git
4. Build with maven: `mvn package`
5. Deploy resulting war https://github.com/matthiasblaesing/SoapHandlerEJBAndServlet.git into started server with context root SoapHandlerEJBAndServletServlet

This is the flaky step: If it does not happen immediately undeply and repeat deployment until message is logged:

> [2017-02-10T21:10:31.736+0100] [Payara 4.1] [SEVERE] [AS-WSJSR109IMPL-00106] [javax.enterprise.webservices] [tid: _ThreadID=132 _ThreadName=admin-thread-pool(6)] [timeMillis: 1486757431736] [levelValue: 1000] [[
>   Handler eu.doppel_helix.dev.soaphandlerejbandservlet.TestHandler instance injection failed: Error creating managed object for class: class eu.doppel_helix.dev.soaphandlerejbandservlet.TestHandler]]

6. Open http://localhost:8080/Endpoint1EjbService/Endpoint1Ejb?Tester and run method

Only "Dependency2#dummy called" but not the "Handler was called" from the soap handler

7. Open http://localhost:8080/SoapHandlerEJBAndServlet/api?Tester and run method

"Dependency1#dummy called" is logged to verify EJB injection works in principle

## Expected Outcome

In step 5: The SoapHandler should always be correctly instantiated
In step 6: "Dependency2#dummy called" _and_ "Handler was called" should always be called

## Current Outcome

In step 5: The SoapHandler is not instantiated/configured and
In step 6: The SoapHandler is not called when the ejb based webservice endpoint is accessed

### Samples

https://github.com/matthiasblaesing/SoapHandlerEJBAndServlet.git

## Investigation
The error in step 5 can be tracked down into the JSR-109 glassfish module. It is logged in:

org.glassfish.webservices.WsUtil.processConfiguredHandlers(List<WebServiceHandler>, Set<String>)

The suppressed stacktrace:

> Schwerwiegend:   com.sun.enterprise.container.common.spi.util.InjectionException: Error creating managed object for class: class eu.doppel_helix.dev.soaphandlerejbandservlet.TestHandler
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.createManagedObject(InjectionManagerImpl.java:352)
> 	at org.glassfish.webservices.WsUtil.processConfiguredHandlers(WsUtil.java:2015)
> 	at org.glassfish.webservices.WsUtil.configureJAXWSServiceHandlers(WsUtil.java:2050)
> 	at org.glassfish.webservices.EjbRuntimeEndpointInfo.prepareInvocation(EjbRuntimeEndpointInfo.java:260)
> 	at org.glassfish.webservices.EjbRuntimeEndpointInfo.initRuntimeInfo(EjbRuntimeEndpointInfo.java:352)
> 	at org.glassfish.webservices.WebServiceEjbEndpointRegistry.registerEndpoint(WebServiceEjbEndpointRegistry.java:125)
> 	at com.sun.ejb.containers.BaseContainer.initializeHome(BaseContainer.java:1193)
> 	at com.sun.ejb.containers.StatelessSessionContainer.initializeHome(StatelessSessionContainer.java:190)
> 	at com.sun.ejb.containers.StatelessContainerFactory.createContainer(StatelessContainerFactory.java:63)
> 	at org.glassfish.ejb.startup.EjbApplication.loadContainers(EjbApplication.java:224)
> 	at org.glassfish.ejb.startup.EjbDeployer.load(EjbDeployer.java:290)
> 	at org.glassfish.ejb.startup.EjbDeployer.load(EjbDeployer.java:100)
> 	at org.glassfish.internal.data.ModuleInfo.load(ModuleInfo.java:206)
> 	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:314)
> 	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:497)
> 	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:220)
> 	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:487)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at javax.security.auth.Subject.doAs(Subject.java:360)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
> 	at java.security.AccessController.doPrivileged(Native Method)
> 	at javax.security.auth.Subject.doAs(Subject.java:360)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
> 	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
> 	at com.sun.enterprise.v3.admin.AdminAdapter.doCommand(AdminAdapter.java:534)
> 	at com.sun.enterprise.v3.admin.AdminAdapter.onMissingResource(AdminAdapter.java:224)
> 	at org.glassfish.grizzly.http.server.StaticHttpHandlerBase.service(StaticHttpHandlerBase.java:189)
> 	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:466)
> 	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:169)
> 	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
> 	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
> 	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
> 	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
> 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
> 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
> 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
> 	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
> 	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
> 	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:526)
> 	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
> 	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
> 	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
> 	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
> 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
> 	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused by: com.sun.enterprise.container.common.spi.util.InjectionException: Exception attempting to inject Local ejb-ref name=eu.doppel_helix.dev.soaphandlerejbandservlet.Endpoint2Servlet/ejbTest,Local 3.x interface =eu.doppel_helix.dev.soaphandlerejbandservlet.Dependency1 resolved to intra-app EJB Dependency1 in module SoapHandlerEJBAndServlet,ejb-link=Dependency1,lookup=,mappedName=,jndi-name=,refType=Session into class eu.doppel_helix.dev.soaphandlerejbandservlet.TestHandler: Lookup failed for 'java:comp/env/eu.doppel_helix.dev.soaphandlerejbandservlet.Endpoint2Servlet/ejbTest' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming}
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl._inject(InjectionManagerImpl.java:740)
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.inject(InjectionManagerImpl.java:507)
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.injectInstance(InjectionManagerImpl.java:141)
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.injectInstance(InjectionManagerImpl.java:127)
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.createManagedObject(InjectionManagerImpl.java:347)
> 	... 52 more
> Caused by: javax.naming.NamingException: Lookup failed for 'java:comp/env/eu.doppel_helix.dev.soaphandlerejbandservlet.Endpoint2Servlet/ejbTest' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming} [Root exception is javax.naming.NameNotFoundException: No object bound for java:comp/env/eu.doppel_helix.dev.soaphandlerejbandservlet.Endpoint2Servlet/ejbTest [Root exception is java.lang.NullPointerException]]
> 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:491)
> 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:438)
> 	at javax.naming.InitialContext.lookup(InitialContext.java:417)
> 	at javax.naming.InitialContext.lookup(InitialContext.java:417)
> 	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl._inject(InjectionManagerImpl.java:636)
> 	... 56 more
> Caused by: javax.naming.NameNotFoundException: No object bound for java:comp/env/eu.doppel_helix.dev.soaphandlerejbandservlet.Endpoint2Servlet/ejbTest [Root exception is java.lang.NullPointerException]
> 	at com.sun.enterprise.naming.impl.JavaURLContext.lookup(JavaURLContext.java:229)
> 	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:471)
> 	... 60 more
> Caused by: java.lang.NullPointerException
> 	at com.sun.ejb.EJBUtils.resolveEjbRefObject(EJBUtils.java:363)
> 	at com.sun.ejb.EjbNamingReferenceManagerImpl.resolveEjbReference(EjbNamingReferenceManagerImpl.java:200)
> 	at com.sun.enterprise.container.common.impl.ComponentEnvManagerImpl$EjbReferenceProxy.create(ComponentEnvManagerImpl.java:1015)
> 	at com.sun.enterprise.naming.impl.GlassfishNamingManagerImpl.lookup(GlassfishNamingManagerImpl.java:745)
> 	at com.sun.enterprise.naming.impl.GlassfishNaming
> Schwerwiegend:   ManagerImpl.lookup(GlassfishNamingManagerImpl.java:715)
> 	at com.sun.enterprise.naming.impl.JavaURLContext.lookup(JavaURLContext.java:159)
> 	... 61 more

## Environment ##

- **Payara Version**: 4.1.1.164
- **Edition**: Full
- **JDK Version**: OpenJDK Runtime Environment (build 1.8.0_121-8u121-b13-0ubuntu1.16.10.2-b13)
- **Operating System**: Ubuntu Yakkety

The problem was first observed on Windows 7, Oracle JDK 8u121, Payara 4.1.1.163.1

--- a/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
+++ b/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
@@ -65,6 +65,7 @@ import org.glassfish.api.event.EventListener;
 import org.glassfish.api.event.Events;
 import org.glassfish.ejb.api.EjbEndpointFacade;
 import org.glassfish.ejb.spi.WSEjbEndpointRegistry;
+import org.glassfish.internal.data.ApplicationInfo;
 import org.glassfish.internal.deployment.Deployment;
 import org.glassfish.webservices.monitoring.WebServiceEngineImpl;
 import org.jvnet.hk2.annotations.Service;
@@ -104,8 +105,10 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
         public void event(EventListener.Event event) {
             if(event.is(Deployment.APPLICATION_LOADED)) {
                 try {
+                    String loadedAppName = ((ApplicationInfo)event.hook()).getName();
                     for(Map.Entry<String, EjbRuntimeEndpointInfo> endpoint : webServiceEjbEndpoints.entrySet()) {
-                        if(!hasMappingFileUri(endpoint.getValue().getEndpoint())) {
+                        String endpointAppName = endpoint.getValue().getEndpoint().getEjbComponentImpl().getEjbBundleDescriptor().getName();
+                        if(loadedAppName.equals(endpointAppName) && !hasMappingFileUri(endpoint.getValue().getEndpoint())) {
                             endpoint.getValue().initRuntimeInfo(adapterListMap.get(endpoint.getKey()));
                         }
                     }

--- a/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
+++ b/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
@@ -37,7 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
-// Portions Copyright [2016] [Payara Foundation]
+// Portions Copyright [2016-2017] [Payara Foundation and/or its affiliates]
 
 package org.glassfish.webservices;
 
@@ -55,11 +55,17 @@ import java.util.StringTokenizer;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
 import javax.inject.Singleton;
 import javax.xml.ws.WebServiceException;
 import javax.xml.ws.handler.Handler;
+import org.glassfish.api.event.EventListener;
+import org.glassfish.api.event.Events;
 import org.glassfish.ejb.api.EjbEndpointFacade;
 import org.glassfish.ejb.spi.WSEjbEndpointRegistry;
+import org.glassfish.internal.deployment.Deployment;
 import org.glassfish.webservices.monitoring.WebServiceEngineImpl;
 import org.jvnet.hk2.annotations.Service;
 
@@ -85,12 +91,40 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
     // root, but that context root must not be used by any web application.  
     // So if the context root portion of the request is in this set, we know
     // the call is for an ejb.
-    private Set<String> ejbContextRoots = new HashSet<String>();
-    
+    private Set<String> ejbContextRoots = new HashSet<>();
 
     // This keeps the list for each service
-    private final Map<String, ServletAdapterList> adapterListMap = new HashMap<String, ServletAdapterList>();
+    private final Map<String, ServletAdapterList> adapterListMap = new HashMap<>();
 
+    private @Inject Events eventService;
+
+    // initialize runtime EJBs after all of it's dependencies have been loaded
+    private final EventListener el = new EventListener() {
+        @Override
+        public void event(EventListener.Event event) {
+            if(event.is(Deployment.APPLICATION_LOADED)) {
+                try {
+                    for(Map.Entry<String, EjbRuntimeEndpointInfo> endpoint : webServiceEjbEndpoints.entrySet()) {
+                        if(!hasMappingFileUri(endpoint.getValue().getEndpoint())) {
+                            endpoint.getValue().initRuntimeInfo(adapterListMap.get(endpoint.getKey()));
+                        }
+                    }
+                } catch (Exception e) {
+                    logger.log(Level.WARNING, LogUtils.EJB_POSTPROCESSING_ERROR, e);
+                }
+            }
+        }
+    };
+
+    @PostConstruct
+    void init() {
+        eventService.register(el);
+    }
+
+    @PreDestroy
+    void destroy() {
+        eventService.unregister(el);
+    }
 
     @Override
     public void registerEndpoint(WebServiceEndpoint webserviceEndpoint,
@@ -117,15 +151,10 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
 
         // notify monitoring layers that a new endpoint is being created.
         WebServiceEngineImpl engine = WebServiceEngineImpl.getInstance();
-        if (endpoint.getEndpoint().getWebService().getMappingFileUri()!=null) {
+        if (hasMappingFileUri(endpoint.getEndpoint())) {
             engine.createHandler((com.sun.xml.rpc.spi.runtime.SystemHandlerDelegate)null, endpoint.getEndpoint());
         } else {
             engine.createHandler(endpoint.getEndpoint());
-            try {
-                endpoint.initRuntimeInfo(adapterListMap.get(uri));
-            } catch (Exception e) {
-                logger.log(Level.WARNING,LogUtils.EJB_POSTPROCESSING_ERROR, e);
-            }
         }
     }
 
@@ -176,7 +205,6 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
         WebServiceEngineImpl engine = WebServiceEngineImpl.getInstance();
 
         engine.removeHandler(endpoint.getEndpoint());
-
     }
     
     /**
@@ -254,5 +282,9 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
             }
             ejbContextRoots = contextRoots;
         }
+    }
+
+    private static boolean hasMappingFileUri(WebServiceEndpoint endpoint) {
+        return endpoint.getWebService().getMappingFileUri() != null;
     }
 }

--- a/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
+++ b/appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServiceEjbEndpointRegistry.java
@@ -43,6 +43,7 @@ package org.glassfish.webservices;
 
 import com.sun.enterprise.container.common.spi.util.InjectionException;
 import com.sun.enterprise.container.common.spi.util.InjectionManager;
+import com.sun.enterprise.deployment.Application;
 import com.sun.enterprise.deployment.WebServiceEndpoint;
 import com.sun.xml.ws.transport.http.servlet.ServletAdapter;
 import com.sun.xml.ws.transport.http.servlet.ServletAdapterList;
@@ -105,7 +106,7 @@ public class WebServiceEjbEndpointRegistry implements WSEjbEndpointRegistry {
         public void event(EventListener.Event event) {
             if(event.is(Deployment.APPLICATION_LOADED)) {
                 try {
-                    String loadedAppName = ((ApplicationInfo)event.hook()).getName();
+                    String loadedAppName = ((ApplicationInfo)event.hook()).getMetaData(Application.class).getName();
                     for(Map.Entry<String, EjbRuntimeEndpointInfo> endpoint : webServiceEjbEndpoints.entrySet()) {
                         String endpointAppName = endpoint.getValue().getEndpoint().getEjbComponentImpl().getEjbBundleDescriptor().getName();
                         if(loadedAppName.equals(endpointAppName) && !hasMappingFileUri(endpoint.getValue().getEndpoint())) {

